File: src/main/java/ini/cx3d/parallelSpatialOrganization/TriangleHashkey.java
/*
Copyright (C) 2009 Frédéric Zubler, Rodney J. Douglas,
Dennis Göhlsdorf, Toby Weston, Andreas Hauri, Roman Bauer,
Sabina Pfister & Adrian M. Whatley.

This file is part of CX3D.

CX3D is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

CX3D is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with CX3D.  If not, see <http://www.gnu.org/licenses/>.
*/

package ini.cx3d.parallelSpatialOrganization;

/**
 * Class to provide hash values for triangles.
 * A simple algorithm is used to calculate a hash value for a triplet of nodes.
 * The calculated hash values do not depend on the order of nodes. Therefore, triangles can
 * be reliably found back even when they were initialized with the same endpoints in a different order.  
 * @author Dennis Goehlsdorf
 *
 * @param <T> The type of user objects associated with nodes in the triangulation.
 */
class TriangleHashKey<T> {
	SpaceNode<T> a, b, c;
	int hashCode;

	/**
	 * Creates a hash value for a triplet of nodes.
	 * @param a The first node.
	 * @param b The second node.
	 * @param c The third node.
	 */
	TriangleHashKey(SpaceNode<T> a, SpaceNode<T> b, SpaceNode<T> c) {
		this.a = a;
		this.b = b;
		this.c = c;
		createHashCode((a != null) ? a.getId() : -1,
				(b != null) ? b.getId() : -1,
				(c != null) ? c.getId() : -1);
	}

	/**
	 * Internal function to create the hash value. Three integer values are needed to compute this value.
	 * @param aId The first node ID.
	 * @param bId The second node ID.
	 * @param cId The third node ID.
	 */
	private void createHashCode(int aId, int bId,
			int cId) {
		hashCode = (Math.min(aId, Math.min(bId, cId))
				* 31
				+ Math.max(aId, Math.max(bId, cId))
				* 11 + aId + bId + cId) % 2000000001;
	}

	/**
	 * @return The hash value associated with this triplet of nodes.
	 */
	public int hashCode() {
		return hashCode;
	}

	/** 
	 * Determines whether this object is equal to another one.
	 * @param obj The object with which this <code>TriangleHashKey</code> should be compared.
	 * @return <code>true</code>, iff <code>obj</code> is an instance of <code>TriangleHashKey</code> and 
	 * 		   refers to the same three points as this <code>TriangleHashKey</code>. 	
	 */
	public boolean equals(Object obj) {
		if (obj instanceof TriangleHashKey) {
			TriangleHashKey<T> other = (TriangleHashKey<T>) obj;
			return ((a == other.a) && (((b == other.b) && (c == other.c)) || ((b == other.c) && (c == other.b))))
					|| ((a == other.b) && (((b == other.a) && (c == other.c)) || ((b == other.c) && (c == other.a))))
					|| ((a == other.c) && (((b == other.a) && (c == other.b)) || ((b == other.b) && (c == other.a))));
		} else
			return false;
	}
}

File: src/spatial_organization/triangle_hash_key.cc
#include "spatial_organization/triangle_hash_key.h"

#include <cmath>

#include "physics/physical_node.h"
#include "spatial_organization/space_node.h"

namespace cx3d {
namespace spatial_organization {

template<class T>
TriangleHashKey<T>::TriangleHashKey(SpaceNode<T>* a,
                                    SpaceNode<T>* b,
                                    SpaceNode<T>* c)
    : a_(a),
      b_(b),
      c_(c),
      hash_code_(0) {
  int a_id = a_ != nullptr ? a_->getId() : -1;
  int b_id = b_ != nullptr ? b_->getId() : -1;
  int c_id = c_ != nullptr ? c_->getId() : -1;
  createHashCode(a_id, b_id, c_id);
}

template<class T>
TriangleHashKey<T>::TriangleHashKey(const TriangleHashKey& other) {
    hash_code_ = other.hash_code_;
    a_ = other.a_;
    b_ = other.b_;
    c_ = other.c_;
}

template<class T>
TriangleHashKey<T>::~TriangleHashKey() {
}

template<class T>
int TriangleHashKey<T>::hashCode() const {
  return hash_code_;
}

template<class T>
bool TriangleHashKey<T>::equalTo(const std::shared_ptr<TriangleHashKey<T>>& other) const {
  return (a_ == other->a_
      && ((b_ == other->b_ && c_ == other->c_) || (b_ == other->c_ && c_ == other->b_)))
      || (a_ == other->b_
          && ((b_ == other->a_ && c_ == other->c_) || (b_ == other->c_ && c_ == other->a_)))
      || (a_ == other->c_
          && ((b_ == other->a_ && c_ == other->b_) || (b_ == other->b_ && c_ == other->a_)));
}

template<class T>
void TriangleHashKey<T>::createHashCode(int a_id, int b_id, int c_id) {
  int min = std::min(a_id, std::min(b_id, c_id));
  int max = std::max(a_id, std::max(b_id, c_id));
  hash_code_ = (min * 31 + max * 11 + a_id + b_id + c_id) % 2000000001;
}

template<class T>
std::size_t TriangleHashKeyHash<T>::operator()(const TriangleHashKey<T>& key) const {
  return key.hash_code_;
}

template<class T>
bool TriangleHashKeyEqual<T>::operator()(const TriangleHashKey<T>& lhs,
                                         const TriangleHashKey<T>& rhs) const {
  return (lhs.a_ == rhs.a_
      && ((lhs.b_ == rhs.b_ && lhs.c_ == rhs.c_)
          || (lhs.b_ == rhs.c_ && lhs.c_ == rhs.b_)))
      || (lhs.a_ == rhs.b_
          && ((lhs.b_ == rhs.a_ && lhs.c_ == rhs.c_)
              || (lhs.b_ == rhs.c_ && lhs.c_ == rhs.a_)))
      || (lhs.a_ == rhs.c_
          && ((lhs.b_ == rhs.a_ && lhs.c_ == rhs.b_)
              || (lhs.b_ == rhs.b_ && lhs.c_ == rhs.a_)));
}

template class TriangleHashKey<cx3d::physics::PhysicalNode>;
template struct TriangleHashKeyHash<cx3d::physics::PhysicalNode>;
template struct TriangleHashKeyEqual<cx3d::physics::PhysicalNode>;

}  // namespace spatial_organization
}  // namespace cx3d

